packages/components/ng-log-to-customer/src/controller.js

```diff
@@ -0,0 +1,168 @@
+    this.setUrlTimeout = null;
+
+    this.setUrlSource().then(() => this.startLog());
+  }
+
+  $onDestroy() {
+    if (this.setUrlTimeout) {
+      this.$timeout.cancel(this.setUrlTimeout);
+    }
+
+    if (this.interval) {
+      this.stop();
+    }
+  }
+
+  handleSearch() {
+    this.logs = this.searchInLogs(this.logs);
+  }
+
+  searchInLogs(logs) {
+    return logs.map((log) => {
+      const highlightedLog = { ...log };
+
+      this.keys.forEach((key) => {
+        highlightedLog[`${key}_highlighted`] = this.safeHighlight(
+          highlightedLog[key].toString(),
+          this.searchText,
+        );
+      });
+
+      return highlightedLog;
+    });
+  }
+
+  safeHighlight(text, phrase) {
+    if (!phrase) return this.$sce.trustAsHtml(text);
+    const highlight = text.replace(
+      new RegExp(`(${phrase})`, 'gi'),
+      '<span class="highlight">$1</span>',
+    );
+    return this.$sce.trustAsHtml(highlight);
+  }
+
+  goToBottom() {
+    this.glued = false;
+    this.$timeout(() => {
+      this.glued = true;
+    });
+  }
+
+  toggleLogs() {
+    if (this.interval) {
+      this.stop();
+    } else {
+      this.startLog();
+    }
+  }
+
+  stop() {
+    this.$interval.cancel(this.interval);
+    this.interval = null;
+  }
+
+  clearSession() {
+    this.logs.length = 0;
+  }
+
+  startLog() {
+    this.interval = this.$interval(() => {
+      this.LogToCustomer.getLogs(this.url).then((data) => {
+        this.appendLogs(data);
+      });
+    }, 3000);
+  }
+
+  appendLogs(data) {
+    if (!this.interval) {
+      // In case user pressed pause with a pending API call, do not append the last received logs
+      return;
+    }
+
+    const logs = this.searchInLogs(this.formatLogs(data));
+
+    this.logs = uniqBy(
+      flatten([...this.logs, logs]),
+      (log) => log._id, // eslint-disable-line no-underscore-dangle
+    );
+  }
+
+  static tryParseJSON(value) {
+    try {
+      return JSON.parse(value);
+    } catch (e) {
+      return value;
+    }
+  }
+
+  getValueByPath(data, pathString) {
+    const pathSegments = pathString.split('.');
+    return pathSegments.reduce((currentValue, segment) => {
+      if (currentValue === undefined) {
+        return undefined;
+      }
+      return this.constructor.tryParseJSON(currentValue[segment]);
+    }, data);
+  }
+
+  formatLogs(data) {
+    return data.messages.map((log) => {
+      return this.keys.reduce((customLog, key) => {
+        const path = this.logKeys[key];
+        // eslint-disable-next-line no-param-reassign
+        customLog[key] = this.getValueByPath(log, path) || '';
+        return customLog;
+      }, {});
+    });
+  }
+
+  setUrlSource() {
+    return this.LogToCustomer.getLogSourceUrl(this.source)
+      .then((data) => {
+        this.url = data.url;
+
+        this.setUrlTimeout = this.$timeout(() => {
+          this.setUrlSource();
+        }, Date.parse(data.expirationDate) - new Date().getTime() - 300);
```

Comment:
Set timeout value in a variable or a function to render it more readable and easier to understand

`const expirationConnectionTimeout = Date.parse(data.expirationDate) - new Date().getTime() - 300)`